File: 3. Longest Substring Without Repeating Characters/solution.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
    const set = new Set();
    let longest = 0;
    let current = 0;
    let subIdx = 0;

    for (let r of s) {
        if (set.has(r)) {
            if (current > longest) {
                longest = current;
            }

            for (; subIdx < s.length; subIdx++) {
                set.delete(s[subIdx]);
                current--;
                if (s[subIdx] == r) {
                    break;
                }
            }

            subIdx++;
        }

        current++;
        set.add(r);
    }

    if (current > longest) {
        longest = current;
    }

    return longest;
};

File: 3. Longest Substring Without Repeating Characters/solution.go
func lengthOfLongestSubstring(s string) int {
	set := map[rune]struct{}{}
	longest := 0
	current := 0
	subIdx := -1
	runes := []rune(s)

	for idx := range runes {
		r := runes[idx]
		if _, ok := set[r]; ok {
			if current > longest {
				longest = current
			}

			for {
				subIdx++
				delete(set, runes[subIdx])
				current--
				if runes[subIdx] == r {
					break
				}
			}
		}

		current++
		set[r] = struct{}{}
	}

	if current > longest {
		longest = current
	}

	return longest
}
